### ISSUE TITLE
Dependency to non-existent header via __has_include always triggers full rebuild in ninja

### ISSUE THREAD

**Issue by ingomueller-net** [llvm/llvm-project#62710]
I am compiling LLVM with Ninja and Clang and noticed that basically all my builds are full builds. Upon investigating, I realized that basically all targets depend on `/usr/include/tbb/tbb.h`, which hasn't been present on my system. No file of LLVM depends on that file *directly*, however, many depend on `/usr/include/x86_64-linux-gnu/c++/12/bits/c++config.h`, which contains the following line:

```cpp
#  define _GLIBCXX_USE_TBB_PAR_BACKEND __has_include(<tbb/tbb.h>)
```

If I remove that line or create an empty file at the location of the missing `tbb.h`, I get partial builds as expected.

One question is whether this is intended behavior or not. On the one hand side, I guess that partial builds should still work even if `__has_include` dependencies are not present. On the other hand side, it was said [in the Ninja mailing list](https://groups.google.com/g/ninja-build/c/O1UDnbMDbVs/m/pjlRL4nJBAAJ) that not reporting a dependency to a non-existing file would mean that a real dependency is missed once that file gets created.

Another one is where this should be reported/discussed/fixed.

Some more details about my set-up below. Let me know if I should report anything else.

- Debian clang version 14.0.6
- libstdc++-12-dev:amd64 version 12.2.0-14
- Ninja version 1.11.1

**Comment by joker-eph** [comment]
Did you find a solution for this? I'm just hitting this now as well.

**Comment by joker-eph** [comment]
Actually wiping out my build directory and reconfiguring fixed the problem.

**Comment by ingomueller-net** [comment]
No, I didn't. I moved on to a different project soon after and only came back to LLVM development later. The problem didn't reoccur and I didn't even remember it. I don't know what made it go away. From the dependencies above, the only things that changed are that my systems upgraded to libstdc++ version 13, and, of course, the LLVM repo changed...
